### PR TITLE
Add DynAlgorithm bindings to python

### PR DESCRIPTION
**Pull request by bernlu**
This PR adds python bindings for the `DynAlgorithm` class in a new `dynalgorithm.pxd`/`dynalgorithm.pyx` file and updates all existing dynamic algorithms to use this base class.
Previously, all python bindings of dynamic algorithms implement the same `update` and `updateBatch` forwarding code, which is now moved to the `DynAlgorithm` base class.

The setup is a little weird because of Cython limitations. First, a figure of the new class hierarchy, where `MyAlgorithm` is any dynamic algorithm:
```
      _CythonParentClass
       ^              ^
       |              |
Algorithm          _CythonSubclassDynAlgorithm
       ^                     ^
       |                     |
MyAlgorithm    -->    DynAlgorithm
```

Explanation:
- Cython (or Python in general) does not allow multiple inheritance of `cdef class`es (also called 'extension types'). Since we use multiple inheritance for dynamic algorithms in c++ (inherit from both `Algorithm` and `DynAlgorithm`), we do need this functionality though. The workaround is to define a wrapper class (`DynAlgorithm`) in pure python which wraps the cdef class. Then we can inherit from `Algorithm` and the pure python wrapper.
- Cython also requires that all classes in a inheritance hierarchy have a common base class. Since there is no such class in C++ (and adding one would probably decrease performance because of more vtables), we instead add a dummy base class just for Cython (`CythonParentClass`). We move the `_Algorithm` pointer to this class s.t. both the `Algorithm` and the `DynAlgorithm` subclass trees have access to that pointer.

In the `_CythonSubclassDynAlgorithm.update` and `updateBatch` methods, we have to use a `dynamic_cast` call to cast the `_Algorithm` pointer to a `_DynAlgorithm` pointer. Regarding performance, I would not expect users to call this often enough to matter, but I also did a quick test and the difference seems to be around 100ns for a single call of dynamic_cast in this context.

With this new setup, implementation of a new dynamic algorithm `MyAlgorithm` now requires less boilerplate and repeated code: 
```py
from .base cimport _Algorithm, Algorithm
from .dynbase cimport _DynAlgorithm
from .dynbase import DynAlgorithm

cdef extern from "<MyAlgorithm.hpp>":
	cdef cppclass _MyAlgorithm "NetworKit::MyAlgorithm"(_Algorithm, _DynAlgorithm):
    # ...

cdef class MyAlgorithm(Algorithm, DynAlgorithm):
    # ... methods specific to MyAlgorithm
    # run, update, updateBatch are inherited from Algorithm and DynAlgorithm
```

### CODE REVIEW

**Comment by fabratu** [comment]
As additional information. Importing the necessary `cppclass _DynAlgorithm` and `class DynAlgorithm` now has to be split (as seen on the PR changes):

```
from .dynbase cimport _DynAlgorithm
from .dynbase cimport DynAlgorithm
...
```

We found no way, creating a workaround using only `cimport`.

File: networkit/dynbase.pyx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+from .dynamics cimport _GraphEvent, GraphEvent
+from .base cimport _CythonParentClass
+from typing import List
+
+ctypedef _DynAlgorithm* _DynAlgorithmPtr
+
+cdef class _CythonSubclassDynAlgorithm(_CythonParentClass):
+	""" Abstract base class for dynamic algorithms
+    Cython does not allow direct inheritance from two extension types (cdef classes).
```
Comment (fabratu) [inline networkit/dynbase.pyx:13]: Indentation for the the docstring looks of. Also move the closing `"""` to another line.

File: networkit/dynbase.pyx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		update(ev)
+
+		The generic update method for updating data structure after an update.
+
+		Parameters
+		----------
+		ev : networkit.dynamics.GraphEvent
+			A graph event.
+
+        Returns
```
Comment (fabratu) [inline networkit/dynbase.pyx:38]: Indentation error.

File: networkit/dynbase.pyx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		updateBatch(batch)
+
+		The generic update method for updating data structure after a batch of updates.
+
+		Parameters
+		----------
+		batch : list(networkit.dynamics.GraphEvent)
+			List of graph events.
+
+        Returns
```
Comment (fabratu) [inline networkit/dynbase.pyx:61]: As above.

File: networkit/base.pyx [hunk 4]
```diff
@@ -1,6 +1,23 @@
 # distutils: language=c++
 
-cdef class Algorithm:
+
+cdef class _CythonParentClass:
+	""" Abstract base class for Cython 
+	The purpose of this class is to provide a single combined base cdef class for all classes that we use in this project.
+	Cython cannot handle multiple inheritance without this single common base class."""
```
Comment (fabratu) [inline networkit/base.pyx:7]: Move the closing `"""` to another line.

**Review by fabratu — CHANGES_REQUESTED** [review]

**Review by fabratu — APPROVED** [review]
LGTM.

**Comment by bernlu** [comment]
Small change: in our `*.pyx` files where we define `cdef cppclass _MyAlgorithm` we should not re-declare `update` and `updateBatch` (it suffices to declare them once in `dynbase`).
This is how we handle `run` as well (most of the time).
It seems like Cython can deal these multi-declarations, even when sometimes functions are declared with `nogil` and sometimes without, but for consistency they should be declared only once in base/dynbase. 
Fixed with this new commit.

**Comment by fabratu** [comment]
> Small change: in our `*.pyx` files where we define `cdef cppclass _MyAlgorithm` we should not re-declare `update` and `updateBatch` (it suffices to declare them once in `dynbase`). This is how we handle `run` as well (most of the time). It seems like Cython can deal these multi-declarations, even when sometimes functions are declared with `nogil` and sometimes without, but for consistency they should be declared only once in base/dynbase. Fixed with this new commit.

Right, missed that in the previous review. Only checked, that update/updateBatch from cdef class were gone.

**Pull request merged**